SFU Refactoring

daniel-abramov opened the pull request:
This closes https://github.com/matrix-org/waterfall/issues/39 and does add structured logging, but not the framework for https://github.com/matrix-org/waterfall/issues/50

The main idea of this refactoring was to provide us with a somewhat stable foundation on top of which we could build new features with as much compile-time safety and with some guarantees that we lower race conditions. 

See https://github.com/matrix-org/waterfall/issues/39 for a description of things that were improved and fixed.

Summary: [TODO, write some better explanation and highlights of this refactoring].

A rough diagram of how components are conceptually connected with each other:
![waterfall-refactored](https://user-images.githubusercontent.com/2765698/203607247-4141c2f4-8299-413a-af1a-810a9e82592f.png)

NB: This is still work-in-progress and **must be tested** before merging.

daniel-abramov committed:
refactor: clean up main.go a bit

daniel-abramov committed:
refactor: renaming modules and functions names

Pretty generic work-in-progress state.

daniel-abramov committed:
refactor: define a new skeleton for the project

This is a WIP state that does not work as expected!

daniel-abramov committed:
refactor: create a better structure for the sources

daniel-abramov committed:
refactor: finalize signaling and peer communication

dbkr reviewed (commented):
This is feeling like a nice refactor so far - using channels & messaging passing in favour of locks definitely feels like a step in the right direction and is good to get done while the project is still a fairly manageable size.

daniel-abramov committed:
refactor: define a package for message sink

This allows to generalize the message sink and get rid of a lot of
copy-paste in the handling functions.

Also this moves types to the right modules, so that `peer` is now
completely matrix-unaware module that contains only plain WebRTC logic.

daniel-abramov committed:
general: rename `src` to `pkg`

This seems to be more idiomatic in Go projects.

daniel-abramov committed:
conference: define sane logic for new participants

New participants are now properly handled depending on whether or not
they were in the session. Also, the conference and peer get documented.

daniel-abramov committed:
conference: handle publish messages from peers

daniel-abramov committed:
paer: unsubscribe from obsolete tracks

Once the track become obsolete, we unsubscribe from them.

daniel-abramov committed:
peer: add naive handling of RTCPs

This is a temporary solution.

daniel-abramov committed:
router: remove useles checks for types

It turns out that currently `mautrix` just simply returns a pointer to
an empty structure in this case (which is probably not the behavior that
we want to have!), but because of this there is no point to check the
types.

daniel-abramov committed:
conference: rename peer message sink for clarity

daniel-abramov committed:
conference: ensure a proper conference lifetime

daniel-abramov committed:
refactor(conference): rename matrix message struct

daniel-abramov committed:
conference: inform owner when the conference ended

This is a quick and a bit dirty implementation of the desired logic. We
definitely must improve it in the future!

daniel-abramov committed:
general: remove custom logger

It seems like it does not really make logs appear more readable
unfortunately, especially in docker environment. We may want to switch
to a different structured logging instead.

See https://github.com/matrix-org/waterfall/issues/50

daniel-abramov committed:
logger: skip checking the TTY

Otherwise the logs in docker would be messed up.

daniel-abramov committed:
router: fix a typo for the OnInvite event

daniel-abramov committed:
router: fix a typo on the type of the printf

Otherwise, the address of the event type would be typed.

daniel-abramov committed:
refactor(conference): simplify new participant

daniel-abramov committed:
conference: add additional logging

daniel-abramov committed:
router: ignore messages for unknown conferences

daniel-abramov committed:
signaling: fix wrongly set party ID

daniel-abramov committed:
ice: add an empty user fragment to the ICE

daniel-abramov committed:
peer: use GatheringCompletePromise from Pion

This is just to test if the outcome of ICE negotiation changes. So far
the connection does not get established despite an exchange of ICE
candidates.

daniel-abramov committed:
Revert "peer: use GatheringCompletePromise from Pion"

This reverts commit 9d20fb3a3f64878d2fd1aa2c3cff4e9a69c28c2e.

daniel-abramov committed:
conference: add additional call invite logging

daniel-abramov committed:
conference: fix unsoundness in channel usage

daniel-abramov committed:
signaling: add additional debug logs

daniel-abramov committed:
conference: add `call_id` handling

The `call_id` does seem to be after all not the same as `conf_id`!

daniel-abramov committed:
conference: use SFU device ID on SelectAnswer

daniel-abramov committed:
config: make log level configurable

daniel-abramov committed:
channel: improve documentation and API surface

daniel-abramov committed:
conference: check for nil rtp tracks

daniel-abramov committed:
conference: fix the subscribe/unsubscribe logic

daniel-abramov committed:
peer: fix a potential segfault

SimonBrandner commented on pkg/signaling/client.go:
```diff
@@ -0,0 +1,67 @@
+
+// Starts the Matrix client and connects to the homeserver,
+// Returns only when the sync with Matrix fails.
+func (m *MatrixClient) RunSync(callback func(*event.Event)) {
```
As far as I understand it, this starts syncing in a loop, could the function name indicate that?

SimonBrandner replied:
`RunSync()` makes it sound like it runs sync only once and then stops which is not the case - it starts syncing and keeps doing that until it encounters an error which makes me think `StartSyncing` (or `SyncUntilError`) would be better

SimonBrandner replied:
`RunSyncing` sounds good

daniel-abramov replied:
Ok, renamed!

daniel-abramov committed:
peer: don't fail on certain RTCP write errors

daniel-abramov committed:
peer: use sdpAnswer from crateAnswer

Instead of calling `LocalDescription()` to get it again. It does not
seem to make any difference in our particular case?

daniel-abramov committed:
conference: handle metadata of streams properly

Don't include tracks in the metadata that are not yet published (for
which we don't have any remote streams available). Also, inform about
metadata changes once tracks get published and unpublished.

daniel-abramov committed:
conference: add hacky way to send unknown metadata

This is a temporary measure to fix the screen sharing.

daniel-abramov committed:
conference: update TODOs and FIXMEs

daniel-abramov committed:
conference: attach stream metadata to offers

daniel-abramov committed:
conference: accept the metadata from the DC offer

SimonBrandner commented on pkg/conference/start.go:
```diff
@@ -0,0 +1,64 @@
+	config Config,
+	signaling signaling.MatrixSignaling,
+	conferenceEndNotifier ConferenceEndNotifier,
+	UserID id.UserID,
```
```suggestion
	userID id.UserID,
```

SimonBrandner commented on pkg/conference/start.go:
```diff
@@ -0,0 +1,64 @@
+		logger:         logrus.WithFields(logrus.Fields{"conf_id": confID}),
+	}
+
+	participantID := ParticipantID{UserID: UserID, DeviceID: inviteEvent.DeviceID, CallID: inviteEvent.CallID}
```
I think I'd prefer this to be done outside this method, it feels like it shouldn't be here

daniel-abramov replied:
The creation of the `participantID`?

SimonBrandner commented on pkg/signaling/matrix.go:
```diff
@@ -0,0 +1,148 @@
+	conferenceID string
+}
+
+// Create a new Matrix client that abstarcts outgoing Matrix messages from a given conference.
```
```suggestion
// Create a new Matrix client that abstracts outgoing Matrix messages from a given conference.
```

daniel-abramov committed:
minor: fix typo in a comment in a signaling comment

Co-authored-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner commented on pkg/conference/processor.go:
```diff
@@ -0,0 +1,220 @@
+		})
+
+	case event.SFUOperationUnpublish:
+		participant.logger.Info("Received unpublish over DC")
```
We should be resending the metadata here and removing the track, right?

SimonBrandner replied:
As discussed in our DM, we only need an `Offer` event and can get rid of `Publish` and `Unpublish` in the future

SimonBrandner commented on pkg/conference/processor.go:
```diff
@@ -0,0 +1,220 @@
+
+		// TODO: Clarify the semantics of unpublish.
+	case event.SFUOperationAlive:
+		// FIXME: Handle the heartbeat message here (updating the last timestamp etc).
```
This needs to be fixed before we merge this, I think

daniel-abramov replied:
Ok, done. Ptal: https://github.com/matrix-org/waterfall/pull/52/commits/28296825506ec805edf6c896217efaa2034852c2 (btw, that's how I normally prefer commits to be done, i.e. "atomic" clean commits with a structured format and that do only one thing; ideally PRs should consist of such commits IMHO [like it's done in Linux kernel repository])

SimonBrandner commented on pkg/conference/processor.go:
```diff
@@ -0,0 +1,220 @@
+func (c *Conference) processMatrixMessage(msg MatrixMessage) {
+	switch ev := msg.Content.(type) {
+	case *event.CallInviteEventContent:
+		c.onNewParticipant(msg.Sender, ev)
```
Perhaps it should be named `onInvite()`?

SimonBrandner commented on pkg/peer/webrtc.go:
```diff
@@ -0,0 +1,171 @@
+
+	switch state {
+	case webrtc.ICEConnectionStateFailed, webrtc.ICEConnectionStateDisconnected:
+		// TODO: Ask Simon if we should do it here as in the previous implementation.
```
The previous implementation doesn't handle this afaik, so we don't have to worry about this for now I think

SimonBrandner commented on pkg/peer/webrtc.go:
```diff
@@ -0,0 +1,171 @@
+		// p.notify <- PeerLeftTheCall{sender: p.data}
+	case webrtc.ICEConnectionStateCompleted, webrtc.ICEConnectionStateConnected:
+		// FIXME: Start keep-alive timer over the data channel to check the connecitons that hanged.
+		// p.notify <- PeerJoinedTheCall{sender: p.data}
```
Do we ever send an event to let the peer know about the end of candidates? I think I've seen it somewhere but it should perhaps be here?

SimonBrandner replied:
Ah, yeah, that sounds right, could we get a comment here to explain why we're not sending anything here?

SimonBrandner commented on pkg/conference/matrix.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+// New participant tries to join the conference.
+func (c *Conference) onNewParticipant(participantID ParticipantID, inviteEvent *event.CallInviteEventContent) error {
```
Perhaps `onInvite()` instead

SimonBrandner reviewed (commented):
This is largely looking good to me though I sometimes struggle a bit not to get lost... Perhaps a better file naming would be a bit, some of the file names aren't very descriptive from my POV. 

I am also a little concerned about completely removing the subscriber/publisher model which is very useful for simulcast and possibly other things; it's also very similar to what ion-sfu and livekit does, iirc

I would also like my PR to get merged before this and for us to somehow integrate it into this one

All of the TODO are also a little concerning

daniel-abramov committed:
Update the documentation in a conference state

Co-authored-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner commented:
> @SimonBrandner,
> 
> > This is largely looking good to me though I sometimes struggle a bit not to get lost... Perhaps a better file naming would be a bit, some of the file names aren't very descriptive from my POV.
> 
> Is there something that could be improved? I want to understand what makes you feel lost. Or is it just because the PR is too large?

The PR being large is definitely a part of it but I feel like the names don't tell me much in some cases...

> I'm definitely in favor of smaller PRs and I'm sorry that this PR is now that big: it's as big as it currently is since it's simply a complete rewrite of a previous implementation (I barely touched `main` and profiling things, but the rest was changed drastically).
> 
> Perhaps it would be easier if I published it as a new repository? (since 90% of things anyway changed) Then it would be easier to review it in an editor (or an IDE) as **a new SFU** rather than trying to comprehend how the new changes relate to the previous implementation. (which I agree must be confusing)
> 
> What I normally do for the large PRs is to split it into smaller "atomic" commits where each commit changes a single thing and explains why it was changed: that's the approach I'm in favor of and that I try to use in all projects (I've noticed that in Matrix and Element it's not always followed: many commits just alter the state of a previous one or combine unrelated things in one commit: something that I want to change!), but I think if I could present it not as one big change, but as series of well-separated commits, it would be much easier to understand (I normally rebase commits before a push and make sure they are clear, but unfortunately this time I published it earlier than I planned so that Dave can take a look at it earlier as we spoke about it).

No worries about the PR being large, no reason to create a separate repo, I realize it's a large refactor and so there aren't many alternatives

> > I am also a little concerned about completely removing the subscriber/publisher model which is very useful for simulcast and possibly other things; it's also very similar to what ion-sfu and livekit does, iirc
> 
> Good point, expected you to ask about it :)
> 
> So, the subscriber and publisher model does exist in the current implementation, but it does not have a separate module or file for it since **for our current implementation** it does not seem to be useful.
> 
> I.e. I can create a function in a `Conference` and name it `getSubscribers()` (it's easy to create one), but then it would immediately pose questions like **what exactly the subscriber structure should contain**, i.e. should it be a structure that contains the ID of a subscriber and the list of IDs of other participants to which the subscriber is subscribed to? Or should it rather be the ID of a subscriber (the participant ID) and the list of tracks? Or should it be the list of tracks and the list of participants to which we're subscribed? Etc... Currently, these can't be answered simply because we don't really need **an explicit** entity such as `Subscriber` or `Publisher`.
> 
> We may indeed need it in the future (e.g. when implementing simulcast)! And once we need it, we definitely give it a good thought and introduce it! But adding these entities right now seemed like a "premature optimization" to me (since it would pose the questions listed above) that would complicate our current (pretty trivial and simple) use case.
> 
> As for the other SFUs, I'm only familiar with some internals of Janus and LiveKit. You mentioned LiveKit: yes, they have **a concept** of a subscription and publishing a track (but that concept also we have in our implementation slightly_smiling_face), but they, like we, don't have special files or packages for `subscriber.go` or `publisher.go`, also there are no such structures as `Subscriber` or `Publisher` in their implementation. They do have a more concrete entity though like `SubscribedTrack` or `Downtrack`.
> 
> So I would not say that we don't have a publisher-subscriber model. In fact, we do have it between peers and a conference (peers publish the messages to which the conference is subscribed). And we also know which tracks everyone is subscribed it and which tracks are being published by each participant (see `participant.go`). We just do it different compared to the implementation in `main` branch - I could not inherit the Publisher/Subscriber logic from main since it unfortunately suffers from race conditions in multiple places and the entities are not separated there, so it also has bugs like a double-subscription (e.g. check `Publisher::Subscribe()` in the `main`). So they had to be reworked anyway [unfortunately].
> 
> And I'm pretty sure that with time we'll also have a more elaborated logic for this with more structures that have a publisher or subscriber-specific logic inside.

Alright, let's leave it out for now and re-add it with simulcast

> > I would also like my PR to get merged before this and for us to somehow integrate it into this one
> 
> Would you like me to integrate the logic from the RTCP PR into this PR before merging it into `main`? Or would you prefer to add it to this PR?
> 
> I can do it, but I was wondering if you would like to do it yourself to get a feel of what the rewritten version looks like and how such things are implemented within the new architecture.

Yeah, I actually wanted to have a go at adding it to this PR exactly for that reason :D Might take a bit of time, but I would like to have a look tomorrow

> > All of the TODO are also a little concerning
> 
> Yep, but there are also TODOs and FIXMEs in `main` :) There is only one place (a FIXME) in this PR that talks about a feature that is implemented in `main`, but not here. The rest existed in the previous implementation (it's just some of them were not marked as concerns and did not have a dedicated TODO, like for instance handling the data channel being busy or closing the SFU once `Sync()` ends). I decided to mark these places with TODOs (if it's not critical) and FIXMEs (if it's more critical and must be discussed), so that we don't forget about them.

Sure, can we just check we're not removing any functionality with this refactor (e.g. keep-alive)?

daniel-abramov committed:
minor: rename `UserID` to `userID`

daniel-abramov committed:
peer: implement heartbeat handling for keepalive

Note that we start the keepalive routine once the peer is created. We do
it like this since the keepalive deadline is actually quite high and I
would say that if within that deadline no heartbeat messages were sent,
then we can consider the connection as stalled.

I.e. starting the keepalive timer only once the peer is connected is
like sparing a second that a peer normally needs to establish a
connection?

daniel-abramov committed:
minor: rename `RunSync` -> `RunSyncing()`

SimonBrandner commented on pkg/conference/participant.go:
```diff
@@ -0,0 +1,51 @@
+	peer            *peer.Peer[ParticipantID]
+	remoteSessionID id.SessionID
+	streamMetadata  event.CallSDPStreamMetadata
+	publishedTracks map[event.SFUTrackDescription]*webrtc.TrackLocalStaticRTP
```
Could we not use `event.SFUTrackDescription` here, please? It doesn't feel good that the Matrix event types and internal types being coupled if they aren't exactly the same thing. Also, `event.SFUTrackDescription` optionally  includes resolution, so it just feels wrong to use it as a key here since we'll never want to use resolution in the key

daniel-abramov replied:
In general, I totally agree with it. I think the only reason I used it like this is that it was not clear what exactly we are going to use [as a track identifier](https://github.com/matrix-org/matrix-spec-proposals/pull/3898#discussion_r1038568858), but I agree that having a separate identifier for tracks makes more sense (it would also simplify the adjacent functions that compare things).

SimonBrandner replied:
Yep, let's merge this ASAP - have made https://github.com/matrix-org/waterfall/issues/56. I might actually have a look into that myself if I get a change before you

SimonBrandner committed:
Remove ugly RTCP handling

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Implement RTCP forwarding

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Further refactor code

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Make `minimalPLIInterval` local

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Fix typo

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Handle missing track

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
`ForwardRTCP` -> `RTCPReceived`

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Decapitalize

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Explain `lastPLITimestamp`

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Use `time.Time` as type

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Simplify `lastPLITimestamp`

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

SimonBrandner committed:
Remove leftover

Signed-off-by: Šimon Brandner <simon.bra.ag@gmail.com>

dbkr commented on pkg/common/channel.go:
```diff
@@ -0,0 +1,70 @@
+	// The channel itself.
+	channel chan<- M
+	// Atomic variable that indicates whether the channel is closed.
+	receiverClosed *atomic.Bool
```
My understanding is that this is considered a bit of a code smell in Go, since a sender should never be trying to write to a closed channel. I guess this means we have something where we can't know this and need to check each time?

daniel-abramov replied:
Good that you've noticed it: tbh I did not like the part of writing this particular structure as it felt like it's not very elegant, but unfortunately I did not find a better way to write it.

The problem that I'm trying to solve is to indeed not write to the closed channel, **but** the problem is that from Go's standpoint the channel is not closed since the channels could only be closed in Go from the sender's side, but not from the receiver's side.

A practical example where it happens in our code:
- There is a router in the code: a structure that simply has a map of all conferences running on the SFU. The router receives signaling messages (To-Device messages), checks their `conf_id,` and simply sends the message to the conference that is responsible for this message. So there is a channel between the router and a particular conference.
- Conference is listening to the channel and reacting to the messages. But at some point, all participants leave the conference which means that the conference must be ended. So the conference grouting and main loop end, sending a message to the router telling "Ok, I'm done, you can remove it".

The problem occurs because at the moment the conference is considered ended, the channel remains open (the conference can't close because the conference only holds a receiver part of the channel). And in go, there is no way to check if the channel is alive or if there is someone listening on the channel. So this means that the router might have sent certain messages to the conference expecting that the conference will read them whereas in reality, the conference has stopped listening on the channel by the moment the router gets to know that the conference is dead.

To sum up:
- There is a way in Go to check if the channel is closed by the sender on the receiver's side.
- There is no way in Go to check if there is any listener on the sender's side.
- We need a way to know (on the sender side) if someone is listening to the channel (in order to know if it makes sense to post a message).

That's pretty much the problem that I tried to solve with this logic, i.e. create a wrapper around the channel where I could inform the sender that the receiver is not listening to the changes, so that if the sender tries to send the message to such a channel, it gets the message back and knows that there are no receivers anymore.

dbkr replied:
Thanks for the detailed explanation - I've also found Go is a bit light on actually explaining how you would write correct & non-racy code sticking to the given paradigms.

daniel-abramov replied:
Right! I've also spent some time trying to understand how to do things safely. In Rust, for instance, we always know if there is a sender and if there is a receiver, but in Go, an elegant implementation of this does not work because of garbage collection: the receivers are not deleted immediately after leaving the scope of a function, but only get removed when the GC kicks in which means that the lifetime of the objects is not strictly defined, hence the consequence of us not knowing when the other counterpart of a channel is dead.

dbkr commented on pkg/common/message_sink.go:
```diff
@@ -0,0 +1,61 @@
+	messageSink chan<- Message[SenderType, MessageType]
+	// Atomic variable that indicates whether the message sink is sealed.
+	// This is used to prevent sending messages to a sealed message sink.
+	// The variable is atomic because it may be accessed from multiple goroutines.
```
For clarity's sake, could we define what 'sealed' means here?

daniel-abramov replied:
Good point, will add a comment about that! Perhaps I need to rewrite the comment for this variable altogether.

dbkr commented on pkg/conference/participant.go:
```diff
@@ -0,0 +1,51 @@
+func (p *Participant) sendDataChannelMessage(toSend event.SFUMessage) {
+	jsonToSend, err := json.Marshal(toSend)
+	if err != nil {
+		p.logger.Error("Failed to marshal data channel message")
```
shouldn't this be returning here, otherwise we'll send blank messages?

daniel-abramov replied:
Good catch, thanks! Will be fixed right away.

daniel-abramov committed:
general: resolve leftovers after rebase
